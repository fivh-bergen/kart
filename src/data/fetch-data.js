import * as fs from 'fs/promises';
import * as path from 'path'

export async function getFetchUrl() {
    const file = path.resolve(path.dirname(''), "./src/data/query.overpassql");
    const data = await fs.readFile(file);
    const query =  Buffer.from(data);
    const urlFormatted = new URLSearchParams()
    urlFormatted.append("data", query.toString())

    const url= new URL("api/interpreter","https://overpass-api.de" )
    return url.href + "?" + urlFormatted.toString();
}

const url = await getFetchUrl();

const response = await fetch(url);

const json = await response.json()

// write to file
await fs.writeFile(path.resolve(path.dirname(''), "./src/data/features.json"), JSON.stringify(json, null, 2));