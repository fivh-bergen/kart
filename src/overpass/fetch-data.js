import * as fs from 'fs/promises';
import * as path from 'path'

export async function getFetchUrl(category) {

    let filePath;

    if(category === 'repair') {
        filePath = path.resolve(path.dirname(''), "./src/overpass/queries/repair.overpassql");
    } else if (category === 'rental') {
        filePath = path.resolve(path.dirname(''), "./src/overpass/queries/rental.overpassql");
    } else {
        filePath = path.resolve(path.dirname(''), "./src/overpass/queries/second-hand.overpassql");
    }
    const data = await fs.readFile(filePath);
    const query =  Buffer.from(data);
    const urlFormatted = new URLSearchParams()
    urlFormatted.append("data", query.toString())

    const url= new URL("api/interpreter","https://overpass-api.de" )
    return url.href + "?" + urlFormatted.toString();
}


const categories = ['rental', 'repair', 'second-hand'];

categories.forEach(async category => {
    const url = await getFetchUrl(category);
    const response = await fetch(url);

    const output = await response.text()

    await fs.writeFile(path.resolve(path.dirname(''), `./src/overpass/data/${category}.osm`), output);
})
