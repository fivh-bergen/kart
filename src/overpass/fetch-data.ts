import * as fs from "fs/promises";
import * as path from "path";
import osmtogeojson from "osmtogeojson";

export async function getFetchUrl(
  category: "repair" | "rental" | "second-hand",
): Promise<string> {
  let filePath;

  if (category === "repair") {
    filePath = path.resolve(
      path.dirname(""),
      "./src/overpass/queries/repair.overpassql",
    );
  } else if (category === "rental") {
    filePath = path.resolve(
      path.dirname(""),
      "./src/overpass/queries/rental.overpassql",
    );
  } else {
    filePath = path.resolve(
      path.dirname(""),
      "./src/overpass/queries/second-hand.overpassql",
    );
  }
  const data = await fs.readFile(filePath);
  const query = Buffer.from(data);
  const urlFormatted = new URLSearchParams();
  urlFormatted.append("data", query.toString());

  const url = new URL("api/interpreter", "https://overpass-api.de");
  return url.href + "?" + urlFormatted.toString();
}

const categories = ["rental", "repair", "second-hand"] as const;

categories.forEach(async (category) => {
  const url = await getFetchUrl(category);
  const response = await fetch(url);

  const output = await response.json();

  const geojson = osmtogeojson(output);

  await fs.writeFile(
    path.resolve(path.dirname(""), `./src/overpass/data/${category}.json`),
    JSON.stringify(geojson, null, 2),
  );
});
