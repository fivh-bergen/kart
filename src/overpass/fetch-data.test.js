import { suite, test } from "node:test";
import { getFetchUrl } from "./fetch-data.js";

suite("getFetchUrl function", () => {
  test("should URL read overpassQL query correctly", async (t) => {
    const actual = await getFetchUrl();
    t.assert.snapshot(actual);
  });
});
