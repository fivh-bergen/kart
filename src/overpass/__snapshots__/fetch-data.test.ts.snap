// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFetchUrl function > should URL read overpassQL query correctly 1`] = `"https://overpass-api.de/api/interpreter?data=%0A%5Bout%3Ajson%5D%3B%0A%2F%2F+fetch+area+%E2%80%9Cbergen%E2%80%9D+to+search+in%0Aarea%28id%3A3600404159%29-%3E.searchArea%3B%0A%0A%2F%2F+find+second+hand+shops%0A%28%0A++node%5B%22shop%22%3D%22second_hand%22%5D%28area.searchArea%29%3B%0A++node%5B%22second_hand%22%3D%22yes%22%5D%28area.searchArea%29%3B%0A++node%5B%22second_hand%22%3D%22only%22%5D%28area.searchArea%29%3B%0A++node%5B%22amenity%22%3D%22give_box%22%5D%28area.searchArea%29%3B%0A%29-%3E.secondHand%3B%0A%0A%2F%2F+send+final+output+to+default+result+set+%0A%28%0A++node.secondHand%3B%0A%29%3B%0A%0Aout+geom%3B"`;
